File: app/src/main/java/com/example/myapplication/CalendarAdapter.java
package com.example.myapplication;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class CalendarAdapter extends RecyclerView.Adapter<CalendarAdapter.VH> {

    public interface OnDayClickListener {
        void onDayClick(int day);
    }

    private final List<Integer> cells;
    private final Map<String, DaySummary> dayMap;
    private final int year;
    private final int month;
    private final int todayDay;
    private final OnDayClickListener onDayClickListener;

    public CalendarAdapter(List<Integer> cells, Map<String, DaySummary> dayMap, 
                           int year, int month, int todayDay, OnDayClickListener onDayClickListener) {
        this.cells = cells;
        this.dayMap = dayMap;
        this.year = year;
        this.month = month;
        this.todayDay = todayDay;
        this.onDayClickListener = onDayClickListener;
    }

    public static class VH extends RecyclerView.ViewHolder {
        public final TextView tvDay;
        public final TextView tvIncome;
        public final TextView tvExpense;
        public final TextView tvDebt;

        public VH(View view) {
            super(view);
            tvDay = view.findViewById(R.id.tvDay);
            tvIncome = view.findViewById(R.id.tvIncome);
            tvExpense = view.findViewById(R.id.tvExpense);
            tvDebt = view.findViewById(R.id.tvDebt);
        }
    }

    @NonNull
    @Override
    public VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_calendar_day, parent, false);
        return new VH(view);
    }

    @Override
    public int getItemCount() {
        return cells.size();
    }

    @Override
    public void onBindViewHolder(@NonNull VH holder, int position) {
        Integer day = cells.get(position);

        if (day == null) {
            holder.tvDay.setText("");
            holder.tvIncome.setText("");
            holder.tvExpense.setText("");
            holder.tvDebt.setText("");
            holder.itemView.setBackground(null);
            holder.itemView.setOnClickListener(null);
            return;
        }

        holder.tvDay.setText(String.valueOf(day));

        Context context = holder.itemView.getContext();
        if (day == todayDay) {
            holder.tvDay.setBackgroundResource(R.drawable.bg_toggle_selected);
            holder.tvDay.setTextColor(ContextCompat.getColor(context, R.color.white));
        } else {
            holder.tvDay.setBackground(null);
            holder.tvDay.setTextColor(ContextCompat.getColor(context, R.color.text_primary));
        }

        String key = String.format(Locale.US, "%04d-%02d-%02d", year, month, day);
        DaySummary summary = dayMap.get(key);

        holder.tvIncome.setText(summary != null && summary.getIncome() > 0 ? "+" + (int) summary.getIncome() : "");
        holder.tvExpense.setText(summary != null && summary.getExpense() > 0 ? "-" + (int) summary.getExpense() : "");
        holder.tvDebt.setText(summary != null && summary.getDebt() > 0 ? "~" + (int) summary.getDebt() : "");

        holder.itemView.setOnClickListener(v -> {
            if (onDayClickListener != null) {
                onDayClickListener.onDayClick(day);
            }
        });
    }
}

File: app/src/main/java/com/example/myapplication/CalendarAdapter.kt
package com.example.myapplication

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import java.util.Locale

class CalendarAdapter(
    private val cells: List<Int?>,
    private val dayMap: Map<String, DaySummary>,
    private val year: Int,
    private val month: Int,
    private val todayDay: Int,
    private val onDayClick: (Int) -> Unit
) : RecyclerView.Adapter<CalendarAdapter.VH>() {

    class VH(view: View) : RecyclerView.ViewHolder(view) {
        val tvDay: TextView = view.findViewById(R.id.tvDay)
        val tvIncome: TextView = view.findViewById(R.id.tvIncome)
        val tvExpense: TextView = view.findViewById(R.id.tvExpense)
        val tvDebt: TextView = view.findViewById(R.id.tvDebt)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH =
        VH(LayoutInflater.from(parent.context).inflate(R.layout.item_calendar_day, parent, false))

    override fun getItemCount() = cells.size

    override fun onBindViewHolder(holder: VH, position: Int) {
        val day = cells[position]

        if (day == null) {
            holder.tvDay.text = ""
            holder.tvIncome.text = ""
            holder.tvExpense.text = ""
            holder.tvDebt.text = ""
            holder.itemView.background = null
            holder.itemView.setOnClickListener(null)
            return
        }

        holder.tvDay.text = day.toString()

        val context = holder.itemView.context
        if (day == todayDay) {
            holder.tvDay.setBackgroundResource(R.drawable.bg_toggle_selected)
            holder.tvDay.setTextColor(ContextCompat.getColor(context, R.color.white))
        } else {
            holder.tvDay.background = null
            holder.tvDay.setTextColor(ContextCompat.getColor(context, R.color.text_primary))
        }

        val key = "%04d-%02d-%02d".format(Locale.US, year, month, day)
        val summary = dayMap[key]

        holder.tvIncome.text  = if (summary != null && summary.income > 0) "+${summary.income.toInt()}" else ""
        holder.tvExpense.text = if (summary != null && summary.expense > 0) "-${summary.expense.toInt()}" else ""
        holder.tvDebt.text    = if (summary != null && summary.debt > 0) "~${summary.debt.toInt()}" else ""

        holder.itemView.setOnClickListener { onDayClick(day) }
    }
}
